src/python/lib/ensembl/compara/utils/cigar.py

```diff
@@ -0,0 +1,160 @@
+        cigar = f"{cigar}{matches}M"
+
+    return cigar
+
+
+def get_cigar_array(cigar:str) -> List[Tuple[int, str]]:
+    """Return an array of the cigar line, e.g.: [(34, 'M'), (12, 'D'), ( 5, 'M') ...]
+
+    Params:
+        cigar: cigar line of the aligned sequence (str)
+
+    Returns:
+        cigar array (List[Tuple[int, str]])
+    Raises:
+        ValueError :  if cigar line incorrect
+    """
+    number = ""
+    result = []
+    for c in cigar:
+        if c.isdigit():
+            number = number + c
+        elif not c.isdigit() and number == "":
+            result.append((1, c))
+        elif not c.isdigit() and number != "":
+            result.append((int(number), c))
+            number = ""
+        else:
+            raise ValueError(f"issue with cigar line {cigar}")
+    return result
+
+
+def alignment_to_seq_coordinate(cigar: str, coord: int) -> List[int]:
+    """Covert coordinate from the alignment level to sequence level
+
+    This function will convert a position at the coordinate level to the sequence level. To be able to
+    deal with gasp and give the maximum information the function is return a list of position. If the
+    position at the alignemnt level fall into a gap then the position just before the gap and just after
+    the gap (in the same order) is returned in the list . If the position is on a non gapped part of the
+    sequence then it return the corresponding position at the sequence level in one list of a unic element.
+
+    Exemple:
+
+        V                              V
+    123456789       return      123456789     return
+    ATGC---CG   --> [4,5]       ATGC---CG  --> [5]
+    1234   56                   1234   56
+
+    Params:
+        cigar: cigar line of the aligned sequence (str)
+        coord: coordinate at the alignment level (List[int])
+
+    Returns:
+        coordinate at the unaligned sequence level (int)
+
+    Raises:
+        ValueError :  if coord is  <= 0 and if coord > alignment length
+    """
+
+    if coord < 1:
+        raise ValueError(f"coord need to be > 0 : current value {coord}")
+
+    result = []
+    cigar_array = get_cigar_array(cigar)
+    seq_level_coord = coord
+    align_index = 0
+    for cigar_elem in cigar_array:
+        align_index = align_index + cigar_elem[0]
+        if cigar_elem[1] == "M":
+            if coord <= align_index:
+                # in this case coord is included in the M cigar_elem
+                # so seq_level_coord  has been trim from all the gap and has now the correct coordinate
+                result = [seq_level_coord]
+                break
+        elif cigar_elem[1] == "D":
+            if coord <= align_index:
+                # in this case coord is included in the D cigar_elem
+                # so we need to trim only a few gaps from seq_level_coord  to have the correct coordinate
+                start_gap = align_index - cigar_elem[0]
+                gap_to_remove = coord - start_gap
```

Comment:
IIUC, the `start_gap` here doesn't contain the 1-based alignment position of the first gap site. That would be calculated as follows:
```
                start_gap = align_index - cigar_elem[0] + 1
```
But calculating `start_gap` in that way would require further changes in subsequent lines, so I'd suggest just renaming the variable to `last_match_pos`, since we could think of `align_index - cigar_elem[0]` as being the 1-based alignment position of the last matching site.
```suggestion
                last_match_pos = align_index - cigar_elem[0]
                gap_to_remove = coord - last_match_pos
```
The name `last_match_pos` doesn't quite apply to the edge case where the first CIGAR op is a deletion, but I would argue that the same is true of `start_gap`.